<a href="https://www.kaggle.com/code/darshanprabhu09/n7-sklearn-pipeline-models-prediction-modelling?scriptVersionId=135424274" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: #FF4500; border-bottom: 8px solid WHITE" > TABLE OF CONTENTS<br><div>  
    
* [1 - DATA DESCRIPTION](#1)
    
* [2 - IMORTING LIBRARIES](#2)
    
* [3 - DATA COLLECTION & PREPARATION](#3)
    
* [4 - EXPLORATORY DATA ANALYSIS](#4)
    * [4.1 - TRAINING DATA](#4.1)
    * [4.2 - TESTING DATA](#4.2)
    * [4.3 - CONVERSION TO NUMPY ARRAY](#4.3)
    * [4.4 - SCALING](#4.4)
    
    
* [5 - MODEL BUILDING](#5)
   * [5.1 - LINEAR REGRESSION](#5.1)
      * [OUTPUT](#5.1.1)
      * [MEAN ABSOLUTE ERROR](#5.1.2)
      * [MODEL FITTING](#5.1.3)
    
      * [6 - PREDICTION](#6)
    
      * [7 - EVALUATION FOR LINEAR REGRESSION MODEL](#7)
    
   * [5.2 - KNN MODEL](#5.2)
      * [6.2 - EVALUATION](#6.2)
    
   
   * [5.3 - RANDOM FOREST REGRESSION](#5.3)
     * [5.3.1 - PREDICTION](#5.3.1)
    
* [8 - CONCLUSION](#8)
    

# <a id="1"></a>
#### <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: brown; border-bottom: 8px solid WHITE" >1 -    DATA DESCRIPTION : 
 <br><div>



**The given dataset provides information on various attributes related to housing in different neighborhoods of California. Here is a partial description of the dataset:**

- **longitude**: The longitude coordinates of the housing location.

- **latitude**: The latitude coordinates of the housing location.

- **housing_median_age**: The median age of houses in a specific neighborhood.

- **total_rooms**: The total number of rooms in a housing unit.

- **total_bedrooms**: The total number of bedrooms in a housing unit.

- **population**: The population count in a specific neighborhood.

- **households**: The number of households in a specific neighborhood.

- **median_income**: The median income of households in a specific neighborhood.

- **median_house_value**: The median value of houses in a specific neighborhood.

- **ocean_proximity**: The proximity of a housing unit to the ocean.

The dataset contains additional information such as labels and counts for different ranges of longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, and median_house_value.


<p style="padding: 10px; background-color: #000000; margin: 0; color: #FF4500; font-family: New Times Roman; font-size: 135%; text-align: left; border-radius: 5px; overflow: hidden; font-weight: 500;">
  Steps to be used to build a Machine Learning Model : 
</p>


## <span style="color: #FF4500;"> Steps </span>


<style>
    body {
        font-family: Arial, sans-serif;
        margin: 20px;
        background-color: #f5f5f5;
    }

    h2 {
        text-align: center;
        color: #333;
    }

    ol {
        margin-left: 30px;
    }

    li {
        margin-bottom: 15px;
    }

    h3 {
        color: #555;
    }

    p {
        margin-bottom: 10px;
    }

    code {
        background-color: #f8f8f8;
        padding: 2px 4px;
        font-family: 'Courier New', monospace;
        font-size: 14px;
    }
</style>

## 7 Steps to Building a SKlearn Pipeline Model

1. **Step 1: Import the necessary libraries**
   
   Import the required libraries such as `sklearn.pipeline` and the desired SKlearn model(s) for your pipeline.

2. **Step 2: Prepare the data**
   
   Load and preprocess your data using techniques like data cleaning, feature engineering, and feature scaling also define the data into "Training data" , "Testing Data" the split generally to be used is (80% for training and 20% for the purpose of Testing).

3. **Step 3: Define the pipeline**
   
   Create an instance of the `Pipeline` class and define the sequence of transformations and the final estimator.

4. **Step 4: Specify preprocessing steps**
   
   For each step in the pipeline, specify the preprocessing operation, such as feature scaling or encoding categorical variables.

5. **Step 5: Fit the pipeline**
   
   Fit the pipeline on your training data, which applies all the specified transformations and trains the final estimator.

6. **Step 6: Evaluate the model**
   
   Assess the performance of your model on the test set using appropriate evaluation metrics like accuracy, precision, or mean squared error.

7. **Step 7: Use the model for predictions**
   
   Once the model is trained and evaluated, you can use it to make predictions on new, unseen data by applying the pipeline transformations and using the final estimator's `predict` method.


# <a id="2"></a>
#### <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: brown; border-bottom: 8px solid WHITE" >2 - IMPORTING THE ESSENTIAL LIBRARIES: 
  <br><div>
#### <span style="color: #FF4500;"> The import statement is used to bring in the libraries into your Python program. numpy is typically imported as np, and pandas is commonly imported as pd. These libraries are widely used for data manipulation, analysis, and processing in Python.




In [1]:
import numpy as np 
import pandas as pd
import sklearn 

# <a id="3"></a>
#### <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: brown; border-bottom: 8px solid WHITE" >3 - DATA COLLECTION & PREPARATION.
.  <br><div>
#### <span style="color: #FF4500;">We did break data into "Training set" and "Testing set" for the purpose of training our model and later on for the evaluation of the model.


In [2]:
test_df = pd.read_csv("/kaggle/input/data-of-california-housing-tale-of-train-and-test/california_housing_test.csv")  #testing dataset
train_df = pd.read_csv("/kaggle/input/data-of-california-housing-tale-of-train-and-test/california_housing_train.csv") #training dataset

# <a id="4"></a>
#### <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: brown; border-bottom: 8px solid WHITE" >4 - EXPLORATORY DATA ANALYSIS.  <br><div>
#### <span style="color: #FF4500;">Exploratory Data Analysis (EDA) is a crucial step in the data analysis process. It involves examining and understanding the dataset to gain insights, discover patterns, and identify potential issues or relationships within the data


# <a id = "4.1"></a>
#### <span style="color: #FF4500;">4.1 : EDA OF TRAINING DATA.</span>

In [3]:
# Display the first few rows of the DataFrame
train_df.head()


longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -114.31     34.19                15.0       5612.0          1283.0   
1    -114.47     34.40                19.0       7650.0          1901.0   
2    -114.56     33.69                17.0        720.0           174.0   
3    -114.57     33.64                14.0       1501.0           337.0   
4    -114.57     33.57                20.0       1454.0           326.0   

   population  households  median_income  median_house_value  
0      1015.0       472.0         1.4936             66900.0  
1      1129.0       463.0         1.8200             80100.0  
2       333.0       117.0         1.6509             85700.0  
3       515.0       226.0         3.1917             73400.0  
4       624.0       262.0         1.9250             65500.0

In [4]:
# Display the last 5 rows of the DataFrame

train_df.tail()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
16995    -124.26     40.58                52.0       2217.0           394.0   
16996    -124.27     40.69                36.0       2349.0           528.0   
16997    -124.30     41.84                17.0       2677.0           531.0   
16998    -124.30     41.80                19.0       2672.0           552.0   
16999    -124.35     40.54                52.0       1820.0           300.0   

       population  households  median_income  median_house_value  
16995       907.0       369.0         2.3571            111400.0  
16996      1194.0       465.0         2.5179             79000.0  
16997      1244.0       456.0         3.0313            103600.0  
16998      1298.0       478.0         1.9797             85800.0  
16999       806.0       270.0         3.0147             94600.0

In [5]:
# Check the dimensions of the DataFrame (rows, columns)
shape = train_df.shape
print("The DataFrame of Training data consists of {} rows and {} columns.".format(shape[0], shape[1]))

The DataFrame of Training data consists of 17000 rows and 9 columns.


In [6]:
# Get summary statistics of the numerical columns
train_df.describe()


longitude      latitude  housing_median_age   total_rooms  \
count  17000.000000  17000.000000        17000.000000  17000.000000   
mean    -119.562108     35.625225           28.589353   2643.664412   
std        2.005166      2.137340           12.586937   2179.947071   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.790000     33.930000           18.000000   1462.000000   
50%     -118.490000     34.250000           29.000000   2127.000000   
75%     -118.000000     37.720000           37.000000   3151.250000   
max     -114.310000     41.950000           52.000000  37937.000000   

       total_bedrooms    population    households  median_income  \
count    17000.000000  17000.000000  17000.000000   17000.000000   
mean       539.410824   1429.573941    501.221941       3.883578   
std        421.499452   1147.852959    384.520841       1.908157   
min          1.000000      3.000000      1.000000       0.499900   
25%        297.000000    790.000000    282.000000       2.566375   
50%        434.000000   1167.000000    409.000000       3.544600   
75%        648.250000   1721.000000    605.250000       4.767000   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        17000.000000  
mean        207300.912353  
std         115983.764387  
min          14999.000000  
25%         119400.000000  
50%         180400.000000  
75%         265000.000000  
max         500001.000000

In [7]:
# Check the data types of each column
train_df.dtypes


longitude             float64
latitude              float64
housing_median_age    float64
total_rooms           float64
total_bedrooms        float64
population            float64
households            float64
median_income         float64
median_house_value    float64
dtype: object

In [8]:
# Check for missing values in the dataset
train_df.isnull().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
dtype: int64

<a id = "4.2"></a>
#### <span style="color: #FF4500;">4.2 : EDA OF TESTING DATA.</span>

In [9]:
# Display the first few rows of the DataFrame
test_df.head()


longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.05     37.37                27.0       3885.0           661.0   
1    -118.30     34.26                43.0       1510.0           310.0   
2    -117.81     33.78                27.0       3589.0           507.0   
3    -118.36     33.82                28.0         67.0            15.0   
4    -119.67     36.33                19.0       1241.0           244.0   

   population  households  median_income  median_house_value  
0      1537.0       606.0         6.6085            344700.0  
1       809.0       277.0         3.5990            176500.0  
2      1484.0       495.0         5.7934            270500.0  
3        49.0        11.0         6.1359            330000.0  
4       850.0       237.0         2.9375             81700.0

In [10]:
# Display the last 5 rows of the DataFrame

test_df.tail()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
2995    -119.86     34.42                23.0       1450.0           642.0   
2996    -118.14     34.06                27.0       5257.0          1082.0   
2997    -119.70     36.30                10.0        956.0           201.0   
2998    -117.12     34.10                40.0         96.0            14.0   
2999    -119.63     34.42                42.0       1765.0           263.0   

      population  households  median_income  median_house_value  
2995      1258.0       607.0         1.1790            225000.0  
2996      3496.0      1036.0         3.3906            237200.0  
2997       693.0       220.0         2.2895             62000.0  
2998        46.0        14.0         3.2708            162500.0  
2999       753.0       260.0         8.5608            500001.0

In [11]:
# Check the dimensions of the DataFrame (rows, columns)
shape = train_df.shape
print("The DataFrame of Testing data consists of {} rows and {} columns.".format(shape[0], shape[1]))

The DataFrame of Testing data consists of 17000 rows and 9 columns.


In [12]:
# Get summary statistics of the numerical columns
test_df.describe()


longitude    latitude  housing_median_age   total_rooms  \
count  3000.000000  3000.00000         3000.000000   3000.000000   
mean   -119.589200    35.63539           28.845333   2599.578667   
std       1.994936     2.12967           12.555396   2155.593332   
min    -124.180000    32.56000            1.000000      6.000000   
25%    -121.810000    33.93000           18.000000   1401.000000   
50%    -118.485000    34.27000           29.000000   2106.000000   
75%    -118.020000    37.69000           37.000000   3129.000000   
max    -114.490000    41.92000           52.000000  30450.000000   

       total_bedrooms    population  households  median_income  \
count     3000.000000   3000.000000  3000.00000    3000.000000   
mean       529.950667   1402.798667   489.91200       3.807272   
std        415.654368   1030.543012   365.42271       1.854512   
min          2.000000      5.000000     2.00000       0.499900   
25%        291.000000    780.000000   273.00000       2.544000   
50%        437.000000   1155.000000   409.50000       3.487150   
75%        636.000000   1742.750000   597.25000       4.656475   
max       5419.000000  11935.000000  4930.00000      15.000100   

       median_house_value  
count          3000.00000  
mean         205846.27500  
std          113119.68747  
min           22500.00000  
25%          121200.00000  
50%          177650.00000  
75%          263975.00000  
max          500001.00000

In [13]:
# Check the data types of each column
test_df.dtypes

longitude             float64
latitude              float64
housing_median_age    float64
total_rooms           float64
total_bedrooms        float64
population            float64
households            float64
median_income         float64
median_house_value    float64
dtype: object

In [14]:
# Check the data types of each column
test_df.dtypes

longitude             float64
latitude              float64
housing_median_age    float64
total_rooms           float64
total_bedrooms        float64
population            float64
households            float64
median_income         float64
median_house_value    float64
dtype: object

<a id = "4.3"></a>
#### <span style="color: #FF4500;">4.3) : CONVERTING THIS DATA INTO NUMPY MATRICES.</span>

In [15]:
X_train,Y_train = train_df.to_numpy()[:,:-1],train_df.to_numpy()[:,-1]

print("The shape of the new matrices are for X_train = {} and for Y_train = {}".format(X_train.shape , Y_train.shape)) 

The shape of the new matrices are for X_train = (17000, 8) and for Y_train = (17000,)


In [16]:
X_test,Y_test = test_df.to_numpy()[:,:-1],test_df.to_numpy()[:,-1]
print("The shape of the new matrices are for X_train = {} and for Y_train = {}".format(X_test.shape , Y_test.shape)) 

The shape of the new matrices are for X_train = (3000, 8) and for Y_train = (3000,)


<a id = "4.4"></a>
#### <span style="color: #FF4500;"> 4.4) SCALING THE VALUES </span>

1.) **StandardScaler** : 
StandardScaler is a data preprocessing technique commonly used in machine learning and statistics. It is used to standardize features or variables by transforming them to have zero mean and unit variance.


2.) **MinMaxScaler** : 
MinMaxScaler is another data preprocessing technique commonly used in machine learning. It transforms features or variables by scaling them to a specific range, typically between 0 and 1. 


3.) **FunctionTransformer** :The FunctionTransformer is a class in scikit-learn that allows you to apply custom transformations to your data using a specified function. It is particularly useful when you want to apply a transformation that is not available as a built-in transformer in scikit-learn.

The FunctionTransformer class takes two important parameters:

*func:* 
This parameter expects a callable function that defines the transformation to be applied to the data.

*validate:* 
This parameter controls whether the input data should be validated before applying the transformation.


In [17]:
from sklearn.preprocessing import StandardScaler , MinMaxScaler , FunctionTransformer 
from copy import deepcopy

In [18]:
# Copying only first two columns i.e the Longitude and Latitude after scaling :
std_scaler = StandardScaler().fit(X_train[:, :2])

# Copying all columns besides the first two :
min_max_scaler = MinMaxScaler().fit(X_train[: , 2:])

# a preprocessor function that performs a preprocessing operation on a numpy array X
def preprocessor(X):
    A = np.copy(X)
    A[: , :2] = std_scaler.transform(X[:,:2])
    A[:,  2:] = min_max_scaler.transform(X[ : , 2: ])
    return A

In [19]:
preprocessor(X_test)

array([[-1.24077729e+00,  8.16354338e-01,  5.09803922e-01, ...,
         4.29944785e-02,  9.94902154e-02,  4.21276948e-01],
       [ 6.29446690e-01, -6.38768279e-01,  8.23529412e-01, ...,
         2.25903192e-02,  4.53872718e-02,  2.13728087e-01],
       [ 8.73822623e-01, -8.63353120e-01,  5.09803922e-01, ...,
         4.15090109e-02,  8.12366387e-02,  3.65063930e-01],
       ...,
       [-6.87702626e-02,  3.15717296e-01,  1.76470588e-01, ...,
         1.93391070e-02,  3.60138135e-02,  1.23418987e-01],
       [ 1.21794383e+00, -7.13629892e-01,  7.64705882e-01, ...,
         1.20519073e-03,  2.13780628e-03,  1.91093916e-01],
       [-3.38594150e-02, -5.63906665e-01,  8.03921569e-01, ...,
         2.10207685e-02,  4.25916790e-02,  5.55916470e-01]])

#### <span style="color: #FF4500;"> VERIFYING THE WORKING OF THE FUNCTION. </span>


# <a id="5"></a>
#### <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: brown; border-bottom: 8px solid WHITE" >5 - MODEL BUILDING<br><div>
    

#### <span style="color: #FF4500;"> This is one of the most essential and also a crucial part known as model building we have a lot of models ,However;based on the problem at given hand we are meant to select a model that provides the most optimal soultion to it with minimal error rate in both the Training as well as the Testing data . It is generally followed after cleaning the data in EDA process . Here we will be constructing three different models to check which one fits the data we are using very well .  They are as follows.

In [20]:
preprocess_transformers = FunctionTransformer(preprocessor)
preprocessor

<function __main__.preprocessor(X)>


<a id = "5.1"></a>
## <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: brown; border-bottom: 8px solid WHITE" >5.1 -  LINEAR REGRESSION MODEL :<br><div>
    
### <span style="color: #FF4500;">Linear regression is a statistical modeling technique used to analyze the relationship between a dependent variable and one or more independent variables. It aims to find the best-fitting linear relationship between the input variables and the output variable.In the context of machine learning, linear regression is often used as a predictive model to estimate or predict the value of a continuous target variable based on input features. The model assumes that there is a linear relationship between the independent variables (features) and the dependent variable (target).Linear regression can be extended to handle more complex relationships by including polynomial features or by using techniques such as regularization (e.g., Ridge regression or Lasso regression) to prevent overfitting.



In [21]:
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression

#### <span style="color: #FF4500;"> CREATING PIPELINES. </span>


## *In data science, a pipeline refers to a sequence of data processing steps that are chained together to form a cohesive workflow. It allows for the efficient and automated execution of various tasks involved in data analysis and model development.*



In [22]:
p1 = Pipeline([('Scalar',preprocess_transformers),('Linear Regression',LinearRegression())])

# In the above line of the code you can see we have passed the steps in form of Tuples in our Pipeline.

# <A ID = "5.1.1"></A>
#### <span style="color: #FF4500;"> OUTPUT FOR THE PIPELINE.</span>


In [23]:
p1

Pipeline(steps=[('Scalar',
                 FunctionTransformer(func=<function preprocessor at 0x7cbb637df5b0>)),
                ('Linear Regression', LinearRegression())])

<A ID = "5.1.2"></A>
#### <span style="color: #FF4500;"> CHECKING ERRORS USING MEAN ABSOLUTE ERROR : . </span>


In [24]:
from sklearn.metrics import mean_absolute_error 

<A ID = "5.1.3"></A>
#### <span style="color: #FF4500;"> Now lets define a function such that it can be used to fit data into pipeline </span>



In [25]:
def fit_and_prnt(p , X_train=X_train , Y_train = Y_train , X_test = X_test , Y_test = Y_test):
    #fitting the pipeline :
    p.fit(X_train , Y_train)
    train_preds = p.predict(X_train)
    test_preds = p.predict(X_test)
    print( "Training error : " + str(mean_absolute_error(train_preds , Y_train)))
    print( "Testing error : " + str(mean_absolute_error(test_preds , Y_test)))


# <a id="6"></a>
#### <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: brown; border-bottom: 8px solid WHITE" >6 - PREDICTION
 <br><div>
    

In [26]:
fit_and_prnt(p1)

Training error : 50795.857117863714
Testing error : 50352.228257942894



<p style="padding: 10px; background-color: #000000; margin: 0; color: #FF4500; font-family: New Times Roman; font-size: 135%; text-align: center; border-radius: 5px; overflow: hidden; font-weight: 500;">
                    LINEAR REGRESSION MODEL:<br>
    As it can be seen here the Training and testing error we got for linear model is :
        
        Training error : 50795.857117863714
        Testing error : 50352.228257942894
</p>

<p style="text-align: center;">
    
</p>



# <a id="7"></a>
#### <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: brown; border-bottom: 8px solid WHITE" >7 - EVALUATION
.  <br><div>
    

<p style="text-align: center;">
    
<span style="color: #FF4500;"> Here are some general guidelines for interpreting the training and testing errors:

**Training Error**: The training error represents how well the model fits the training data. In your case, the training error is approximately 50795.86. A lower training error indicates that the model is able to capture the patterns and relationships in the training data more accurately. However, an extremely low training error might indicate overfitting, where the model has memorized the training data instead of learning generalizable patterns.

**Testing Error**: The testing error represents how well the model performs on unseen data, i.e., the testing set. In your case, the testing error is approximately 50352.23. The testing error is typically higher than the training error because the model encounters new data during testing. The goal is to have a testing error that is close to the training error, indicating that the model can generalize well to unseen data.

To assess the performance of your model, you can compare the testing error with other models or with a predefined threshold that determines what is an acceptable level of error for your specific problem. Additionally, you can use other evaluation metrics such as R-squared, mean absolute error (MAE), or domain-specific metrics to get a more comprehensive understanding of the model's performance.

Keep in mind that the evaluation of a model is highly dependent on the specific problem, dataset, and domain knowledge. It's always recommended to analyze the results in context and consider the trade-offs between different evaluation metrics and requirements.: .</span>


## <span style="color: #FF4500;"> Since the Error rate in Training and Testing data was too high we will try changing the model , lets import KNN and look how it will be working.</span>


# <a id="5.2"></a>
## <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: brown; border-bottom: 8px solid WHITE" >5.2 - KNN MODEL {K - NEAREST NEIGHBOUR}  <br><div>
    


### <span style="color: #FF4500;"> K-Nearest Neighbors (KNN) is a non-parametric machine learning algorithm used for both classification and regression tasks. It is a simple and intuitive algorithm that makes predictions based on the similarity of data points.It's worth noting that the choice of distance metric and the value of K can significantly impact the performance of the KNN algorithm. Different distance metrics may be more suitable for different types of data. Additionally, the scalability of KNN can be a challenge as it requires comparing the new data point with all other data points in the training set.KNN is often used as a baseline or in combination with other algorithms for more complex machine learning tasks. It is particularly useful in cases where the decision boundaries are non-linear or the underlying data distribution is unknown.




<p style="padding: 10px; background-color: #000000; margin: 0; color: #FF4500; font-family: New Times Roman; font-size: 135%; text-align: center; border-radius: 5px; overflow: hidden; font-weight: 500;">
                    NOTE :<br>
    STEPS 1 - 6 (DATA COLLECTION , DATA PREPROCESSING AND EDA) IF NOT REPEATED WILL NOT MATTER SINCE WE HAVE ALREADY CLEANED THE DATA ACCORDING TO OUR USE . SO WE WILL BE DIRECTLY HOPPING TO "MODEL BUILIDNG"</p>

<p style="text-align: center;">
    
</p>



In [27]:
from sklearn.neighbors import KNeighborsRegressor as KNR


# Using a second pipeline with the same scaling method but choosing KNN regression instead of Linear regression: 
# Also defining the no. of neighbors as "5"
p2 = Pipeline([('Scalar',preprocess_transformers),("KNN Regression",KNR(n_neighbors = 5))])
fit_and_prnt(p2)

Training error : 28343.652870588237
Testing error : 36045.85766666666


In [28]:
p2

Pipeline(steps=[('Scalar',
                 FunctionTransformer(func=<function preprocessor at 0x7cbb637df5b0>)),
                ('KNN Regression', KNeighborsRegressor())])


<p style="padding: 10px; background-color: #000000; margin: 0; color: #FF4500; font-family: New Times Roman; font-size: 135%; text-align: center; border-radius: 5px; overflow: hidden; font-weight: 500;">
                    KNN MODEL:<br>
    As it can be seen here the Training and testing error we got for KNN model is :
        
        Training error : 28343.652870588237
        Testing error : 36045.85766666666
</p>

<p style="text-align: center;">
    
</p>



# <a id="6.2"></a>
#### <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: brown; border-bottom: 8px solid WHITE" >6.2 - KNN MODEL EVALUATION  <br><div>
    

    
<span style="color: #FF4500;">It's important to interpret these errors within the context of your problem and domain knowledge. Evaluating the model's performance involves considering the trade-offs between different evaluation metrics and the specific requirements of your application.

Training Error: The training error is approximately 28343.65. This indicates how well the model fits the training data. A lower training error suggests that the model is able to capture the patterns and relationships in the training data more accurately.

Testing Error: The testing error is approximately 36045.86. This represents how well the model performs on unseen data, i.e., the testing set. The testing error is typically higher than the training error since the model encounters new data during testing. The goal is to have a testing error that is close to the training error, indicating that the model can generalize well to unseen data. .</span>


# <a id="5.3"></a>
## <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: brown; border-bottom: 8px solid WHITE" >5.3 - RANDOM FOREST REGRESSION.  <br><div>
    

### <span style="color: #FF4500;">RFR stands for Random Forest Regression, which is a supervised machine learning algorithm used for regression tasks. It is an extension of the Random Forest algorithm, which is primarily used for classification.Random Forest Regression combines the concept of ensemble learning with decision tree models to make predictions. It creates an ensemble of multiple decision trees and averages their predictions to obtain a final prediction. Each decision tree is constructed using a randomly selected subset of the training data and a subset of input features, providing diversity and reducing overfitting.Random Forest Regression offers several advantages, including:</span>
    
* It can handle a large number of input features and works well with high-dimensional data.
* It reduces overfitting by combining predictions from multiple trees.
* It provides an estimate of feature importance, allowing for feature selection.
* It can capture non-linear relationships between features and target variables.
    
    
### <span style="color: #FF4500;">However, Random Forest Regression also has some limitations, such as:</span>

* It can be computationally expensive, especially when dealing with a large number of trees or complex datasets.
* It may not perform well when there are strong linear relationships between features and target variables.

### <span style="color: #FF4500;">Overall, Random Forest Regression is a powerful algorithm that is widely used in regression tasks, particularly when dealing with complex and non-linear relationships</span>




In [29]:
from sklearn.ensemble import RandomForestRegressor as RFR


# Using a second pipeline with the same scaling method but choosing Random forest regression instead of KNN: 
# Also defining the no. of estimators  as "10" and depth as "5"
p3 = Pipeline([('Scalar',preprocess_transformers),("Random Forest",RFR(n_estimators = 10 , max_depth = 7))])
fit_and_prnt(p3)


Training error : 41017.425379547356
Testing error : 43904.37611561218


# <a id="5.3.1"></a>
#### <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: brown; border-bottom: 8px solid WHITE" >RFR model prediction.  <br><div>
    As it can be seen here the Training and testing error we got for RFR model is :
        
          Training error : 41583.216967626926
          Testing error : 44229.32230923467


# <a id="8"></a>
## <div style= "font-family: Times New Roman; font-weight:bold; letter-spacing: 0px; color:WHITE; font-size:120%; text-align:left;padding:3.0px; background: brown; border-bottom: 8px solid WHITE" > 8 - CONCLUSION  <br><div>
    

<p style="text-align: center;">
 
# <span style="color: #FF4500;"> THE ERROR RATE IN ALL THREE ABOVE MODELS WHEN COMPARED GIVES :
    
    
    1.) As it can be seen here the Training and testing error we got for linear model is :
        Training error : 50795.857117863714
        Testing error : 50352.228257942894
  
    
    2.) As it can be seen here the Training and testing error we got for KNN model is :
        Training error : 28343.652870588237
        Testing error : 36045.85766666666
    
    
    3.) As it can be seen here the Training and testing error we got for RFR model is :
          Training error : 41583.216967626926
          Testing error : 44229.32230923467
 </span>




<span style="color: #FF4500;"> FROM THE CODE IT CAN BE CONCLUDED THAT KNN HAS LOWEST ERROR RATE IN TERMS OF TRAINING DATASET AS WELL AS TESTING DATASET . 
 .</span>



![Alt Text](https://cdn.dribbble.com/users/448124/screenshots/2177658/qq-.gif)
# <span style="color: skyblue;"> "Thank you for checking out my notebook! If you found it helpful, an upvote would be greatly appreciated.As It helps my work to reach furthermore and spread the notebook more deeply and widely in the community that may help a lot of people to learn about skLearn znd the models I have worked on this project . Happy coding!" . 
 .</span>